xbmc/platform/android/activity/XBMCApp.cpp

```diff
@@ -1621,3 +1624,44 @@ void CXBMCApp::surfaceDestroyed(CJNISurfaceHolder holder)
 
   m_window.reset();
 }
+
+bool CXBMCApp::OnVoiceRecognition()
+{
+  if (CJNISpeechRecognizer::isRecognitionAvailable(CXBMCApp::Get()))
+  {
+    if (CJNIContext::checkCallingOrSelfPermission("android.permission.RECORD_AUDIO") ==
+        CJNIPackageManager::PERMISSION_GRANTED) 
+    {
+      // must be called from the main thread
+      runNativeOnUiThread(StartRecognition, nullptr);
+      return true;
+    }
+    else
+    {
+      CLog::Log(LOGDEBUG, "CXBMCApp OnVoiceRecognition(): permission RECORD_AUDIO is not allowed");
+      return false;
+    }
+  }
+  else
+  {
+    CLog::Log(LOGDEBUG, "CXBMCApp OnVoiceRecognition(): Google speech service no available");
+    return false;
+  }
+}
+
+void CXBMCApp::StartRecognition(CVariant*)
+{
+  {
+    CLog::Log(LOGDEBUG, "CXBMCApp StartRecognition()");
+
+    CJNIIntent intent = CJNIIntent(CJNIRecognizerIntent::ACTION_RECOGNIZE_SPEECH);
+    intent.putExtra(CJNIRecognizerIntent::EXTRA_LANGUAGE_MODEL,
+                    CJNIRecognizerIntent::LANGUAGE_MODEL_FREE_FORM);
+
+    CJNISpeechRecognizer speechRecognizer =
+        CJNISpeechRecognizer::createSpeechRecognizer(CXBMCApp::Get());
+  
+    speechRecognizer.setRecognitionListener(CJNIXBMCRecognitionListener());
+    speechRecognizer.startListening(intent);
```

Comment:
I would order the code  bit so that things that belong logically together are close together in the code.

```c++
     CJNISpeechRecognizer speechRecognizer =
         CJNISpeechRecognizer::createSpeechRecognizer(CXBMCApp::Get());

     speechRecognizer.setRecognitionListener(CJNIXBMCRecognitionListener());
     
     CJNIIntent intent = CJNIIntent(CJNIRecognizerIntent::ACTION_RECOGNIZE_SPEECH);
     intent.putExtra(CJNIRecognizerIntent::EXTRA_LANGUAGE_MODEL,
                     CJNIRecognizerIntent::LANGUAGE_MODEL_FREE_FORM); 
     speechRecognizer.startListening(intent);
```

Reply:
ok, I will reorder the code